### PR TITLE
feat: support wallx model

### PR DESCRIPTION
**Pull request by wut19**
This pull request introduces support for the new Wall-X policy, a Vision-Language-Action model for cross-embodiment robotic control, into the `lerobot` codebase. The changes include adding the Wall-X policy configuration, model, processor, and associated dependencies, as well as updating the policy factory to recognize and instantiate Wall-X. These updates enable users to train and deploy the Wall-X policy using the same unified interface as other policies in the library.

**Wall-X Policy Integration**

* Added the Wall-X policy implementation, including `WallXConfig`, `WallXPolicy`, and processor logic in the new `src/lerobot/policies/wall_x/` module. This includes configuration validation, optimizer/scheduler presets, and pre/post-processing pipelines tailored for Wall-X. [[1]](diffhunk://#diff-19d4d9accdd1b95dd7c443fdea661a27e9926df1fd9a40b573ddf39935b8535dR1-R162) [[2]](diffhunk://#diff-ec2cffa76aea72bb4bd506747376524acdf9c6c5adec51443086d3421b1f50e6R1-R21) [[3]](diffhunk://#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8R1-R135) [[4]](diffhunk://#diff-6fddced46c9ffc7e57ac7b26a9c2e4c80001d46ae30ffe171290d39a07ed4a08R1-R43)
* Added a README for Wall-X describing its capabilities, architecture, and citation information.

**Policy Factory and Registry Updates**

* Registered Wall-X in the policy factory, enabling instantiation and configuration via the standard API (`get_policy_class`, `make_policy_config`, and `make_pre_post_processors`). Factory docstrings updated to reflect Wall-X as a supported policy type. [[1]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35R45) [[2]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35L64-R65) [[3]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35R117-R120) [[4]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35L133-R138) [[5]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35R169-R170) [[6]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35R354) [[7]](diffhunk://#diff-5a185e5876e964030c0f774873f168bf2e595d82b1ce544c27990e3bd6848d35R365-R372)
* Updated the policy module `__init__.py` to include `WallXConfig` in the export list for easier imports and discoverability. [[1]](diffhunk://#diff-11ef5efc4f3538e86d3b9d1b97d188190fb335a875e1b76ad6a431523d919488R25) [[2]](diffhunk://#diff-11ef5efc4f3538e86d3b9d1b97d188190fb335a875e1b76ad6a431523d919488R37)

**Dependency Management**

* Added a new optional dependency group `wallx` in `pyproject.toml` with required packages for training and inference (torch, transformers, accelerate, etc.), and included it in the `all` dependency group for convenience. [[1]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R123-R134) [[2]](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R174)

### CODE REVIEW

File: src/lerobot/policies/wall_x/configuration_wall_x.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        default_factory=lambda: {
+            "VISUAL": NormalizationMode.IDENTITY,
+            "STATE": NormalizationMode.MEAN_STD,
+            "ACTION": NormalizationMode.MEAN_STD,
+        }
+    )
+
+    # ==================== Action Prediction ====================    
+    # Pretrained model paths
+    pretrained_name_or_path: str = "x-square-robot/wall-oss-flow"
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/configuration_wall_x.py:54]: Just to confirm, in case I want to use x-square-robot/wall-oss-fast, I just need to change `pretrained_name_or_path`  and `prediction_mode` to `fast and `use_fast_tokenizer` to `True`?

File: src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        self.noise_scheduler = noise_scheduler
+        self.dim_inputs = tuple(dim_inputs)
+        self.attention_moe = attention_moe
+        self.mlp_moe = mlp_moe
+
+        # Validate the correctness of rotary position embeddings parameters
+        # BC: if there is a 'type' field, move it to 'rope_type'.
+        # and change type from 'mrope' to 'default' because `mrope` does defeault RoPE calculations
+        # one can set it to "linear"/"dynamic" etc. to have scaled RoPE
+        # TODO: @raushan update config in the hub
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py]: There is still one TODO here

File: tests/policies/wall_x/__init__.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Wall-X policy tests
```
Comment (pkooij) [inline tests/policies/wall_x/__init__.py]: You can remove this file

File: src/lerobot/policies/wall_x/utils.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+import torch
+from transformers import BatchFeature
+
+from lerobot.policies.wall_x.constant import (
+    CAMERA_NAME_MAPPING,
+)
+from lerobot.utils.constants import OBS_IMAGES
+
+
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/utils.py:38]: The pipeline/processors (pre and post) that we have in LeRobot can be used for tokenization and all preprocessing you do here, for now having these utils is fine but we might refactor this to be different pipeline steps in the future.

File: pyproject.toml [hunk 5]
```diff
@@ -120,6 +120,18 @@ intelrealsense = [
...
+    "torch==2.6.0",
+    "torchvision==0.21.0",
+    "torchaudio==2.6.0",
+    "transformers==4.49.0",
+    "accelerate==1.10.1",
+    "peft==0.17.1",
+    "scipy==1.15.3",
+    "torchdiffeq==0.2.5",
+    "qwen_vl_utils==0.0.11",
+    "flash-attn==2.7.4.post1"
```
Comment (pkooij) [inline pyproject.toml]: It would be better and make wall-oss more accessible if we can choose attn_implementation and not require flash_attention by default, so remove it here and also in WalllOSS config and implementation  add `attn_implementation` as choice

File: pyproject.toml [hunk 6]
```diff
@@ -120,6 +120,18 @@ intelrealsense = [
 phone = ["hebi-py>=2.8.0,<2.12.0", "teleop>=0.1.0,<0.2.0", "fastapi<1.0"]
 
 # Policies
+wallx = [
+    "torch==2.6.0",
+    "torchvision==0.21.0",
+    "torchaudio==2.6.0",
+    "transformers==4.49.0",
```
Comment (pkooij) [inline pyproject.toml:124]: Which torchcodec version are you using?

Comment (wut19) [reply pyproject.toml:124]: We don't require a specific torchcodec version for wallx but follow the latest lerobot requirements. We are using torchcodec==0.5

File: src/lerobot/policies/wall_x/modeling_wall_x.py [hunk 7]
```diff
@@ -0,0 +1,2107 @@
...
+                agent_pos,
+                torch.zeros(agent_pos.shape[0], agent_pos.shape[1], pad_size, device=agent_pos.device)
+            ], dim=-1)
+            agent_pos_mask = torch.cat([
+                agent_pos_mask,
+                torch.zeros(agent_pos_mask.shape[0], agent_pos_mask.shape[1], pad_size, device=agent_pos_mask.device)
+            ], dim=-1)
+        
+        # ==================== PROCESS ACTIONS ====================
+        action = batch[ACTION]  # (batch_size, chunk_size, action_dim)
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/modeling_wall_x.py]: During inference we don't have an actions (this will give key error)  so we have to do action = get(batch[ACTION], None) and handle this case where action is None and create dummy action and mask in this case.

Comment (wut19) [reply src/lerobot/policies/wall_x/modeling_wall_x.py]: We have fixed this in the latest commit.

File: pyproject.toml (hunk 5 above)
Comment (wut19) [reply pyproject.toml]: We have added a param `attn_implementation` in wallx config and removed the flash-attn requirement.

File: src/lerobot/policies/wall_x/configuration_wall_x.py (hunk 1 above)
Comment (wut19) [reply src/lerobot/policies/wall_x/configuration_wall_x.py:54]: We now only need to change `pretrained_name_or_path` and `prediction_mode`, `use_fast_tokenizer` is automatically set.

File: src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+        self.vocab_size = vocab_size
+        self.max_position_embeddings = max_position_embeddings
+        self.hidden_size = hidden_size
+        self.intermediate_size = intermediate_size
+        self.num_hidden_layers = num_hidden_layers
+        self.num_attention_heads = num_attention_heads
+        self.use_sliding_window = use_sliding_window
+        self.sliding_window = sliding_window
+        self.max_window_layers = max_window_layers
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py:211]: Also add `self.layer_types = ["dense"] * num_hidden_layers

File: src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    >>> # Initializing a Qwen2_5_VL style configuration
+    >>> configuration = Qwen2_5_VLConfig()
+
+    >>> # Initializing a model from the Qwen2-VL-7B style configuration
+    >>> model = Qwen2_5_VLForConditionalGeneration(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
```
Comment (pkooij) [inline src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py:148]: We should also add: 
```
@property 
def text_config(self):
   return self
```

File: src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py (hunk 8 above)
Comment (wut19) [reply src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py:211]: We have added this.

File: src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py (hunk 9 above)
Comment (wut19) [reply src/lerobot/policies/wall_x/qwen_model/configuration_qwen2_5_vl.py:148]: We have added this

File: pyproject.toml
[the remark this answers is not in the document]
Comment (wut19) [reply pyproject.toml]: should I modify to `transformers>=4.49.0,<5.0.0` or `transformers==4.49.0`

**Review by pkooij — APPROVED** [review]

**Review by pkooij — DISMISSED** [review]
